### DISCUSSION TITLE
back_populates does not behave as expected when lazy="raise"

### DISCUSSION THREAD

**Discussion by dkbarn** [sqlalchemy/sqlalchemy#11556] [Usage Questions] [answered]
Consider the following simplistic example:
```python
from __future__ import annotations

import logging
from sqlalchemy import create_engine, Integer, ForeignKey, select
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker, Mapped, mapped_column, relationship, joinedload

logging.basicConfig(level=logging.INFO)
logging.getLogger("sqlalchemy.engine").setLevel(logging.INFO)

engine = create_engine("sqlite://")
Session = sessionmaker(bind=engine)
Base = declarative_base()

class Media(Base):
    __tablename__ = "media"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)

    instance: Mapped[MediaInstance | None] = relationship(
        back_populates="media",
        cascade="all, delete-orphan",
        single_parent=True,
        lazy="raise"
    )

class MediaInstance(Base):
    __tablename__ = "media_instance"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    media_id: Mapped[int] = mapped_column(ForeignKey("media.id"))
    media: Mapped[Media] = relationship(
        back_populates="instance",
        lazy="raise",
    )

# create the table
Base.metadata.create_all(engine)

# add some data
with Session() as session:
    with session.begin():
        media = Media()
        instance = MediaInstance(media=media)
        session.add_all([media, instance])

# fetch data
with Session() as session:
    media = session.scalar(
        select(Media)
            .limit(1)
            .options(
                joinedload(Media.instance),
            )
    )
    assert media.instance
    assert media.instance.media
```

When you run this code, the final assertion raises the exception:
```
sqlalchemy.exc.InvalidRequestError: 'MediaInstance.media' is not available due to lazy='raise'
```

My expectation would be that no exception is raised, because back_populates has been defined on both sides of the relationship. Accessing the `MediaInstance.media` relationship will not result in another SQL query being issued, because it has been back-populated via the `Media.instance` relationship.

This is confirmed if you comment out both of the `lazy="raise"` attributes. With this change, there is no exception raised, but also there is no additional SQL query issued to the database. So my question is, why can't back-populates do what it is intended to do even when lazy="raise"? My understanding was that the purpose of lazy="raise" is to protect against any unexpected communication with the database. But there is no further communication with the database required in order to access `media.instance.media`.

**Comment by zzzeek** [comment, marked as answer]
> Consider the following simplistic example:
> 
> ```python
> from __future__ import annotations
> 
> import logging
> from sqlalchemy import create_engine, Integer, ForeignKey, select
> from sqlalchemy.ext.declarative import declarative_base
> from sqlalchemy.orm import sessionmaker, Mapped, mapped_column, relationship, joinedload
> 
> logging.basicConfig(level=logging.INFO)
> logging.getLogger("sqlalchemy.engine").setLevel(logging.INFO)
> 
> engine = create_engine("sqlite://")
> Session = sessionmaker(bind=engine)
> Base = declarative_base()
> 
> class Media(Base):
>     __tablename__ = "media"
> 
>     id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
> 
>     instance: Mapped[MediaInstance | None] = relationship(
>         back_populates="media",
>         cascade="all, delete-orphan",
>         single_parent=True,
>         lazy="raise"
>     )
> 
> class MediaInstance(Base):
>     __tablename__ = "media_instance"
> 
>     id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
>     media_id: Mapped[int] = mapped_column(ForeignKey("media.id"))
>     media: Mapped[Media] = relationship(
>         back_populates="instance",
>         lazy="raise",
>     )
> 
> # create the table
> Base.metadata.create_all(engine)
> 
> # add some data
> with Session() as session:
>     with session.begin():
>         media = Media()
>         instance = MediaInstance(media=media)
>         session.add_all([media, instance])
> 
> # fetch data
> with Session() as session:
>     media = session.scalar(
>         select(Media)
>             .limit(1)
>             .options(
>                 joinedload(Media.instance),
>             )
>     )
>     assert media.instance
>     assert media.instance.media
> ```
> 
> When you run this code, the final assertion raises the exception:
> 
> ```
> sqlalchemy.exc.InvalidRequestError: 'MediaInstance.media' is not available due to lazy='raise'
> ```
> 
> My expectation would be that no exception is raised, because back_populates has been defined on both sides of the relationship. Accessing the `MediaInstance.media` relationship will not result in another SQL query being issued, because it has been back-populated via the `Media.instance` relationship.

Correct, however if you want to block lazy loading at the point of SQL you would use `lazy="raise_on_sql"`.  this allows lazy loads that load the object from the local `Session` without emitting SQL to still proceed.

https://docs.sqlalchemy.org/en/20/tutorial/orm_related_objects.html#raiseload

**Reply by dkbarn** [reply]
Ahh thank you for that clarification. I was not aware that "raise_on_sql" actually has different behaviour than "raise", I thought the two were interchangeable.

https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html#summary-of-relationship-loading-styles
> **raise loading** - available via lazy='raise', lazy='raise_on_sql', or the [raiseload()](https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html#sqlalchemy.orm.raiseload) option, this form of loading is triggered at the same time a lazy load would normally occur, except it raises an ORM exception in order to guard against the application making unwanted lazy loads. An introduction to raise loading is at [Preventing unwanted lazy loads using raiseload](https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html#prevent-lazy-with-raiseload).

This text led to believe that 'raise' and 'raise_on_sql' were identical, because they are mentioned together under the same bullet point, and there is only a single description given for them, with no distinction made between their differing behaviour. It might be worth clarifying this a bit here (I realize the distinction is made elsewhere, in the documentation for raiseload).
